libzfs_core/_libzfs_core.py

```diff
@@ -0,0 +1,683 @@
+# Copyright 2015 ClusterHQ. See LICENSE file for details.
+
+"""
+Python wrappers for libzfs_core interfaces.
+
+As a rule, there is a Python function for each C function.
+The signatures of the Python functions generally follow those of the
+functions, but the argument types are natural to Python.
+nvlists are wrapped as dictionaries or lists depending on their usage.
+Some parameters have default values depending on typical use for
+increased convenience.  Output parameters are not used and return values
+are directly returned.  Error conditions are signalled by exceptions
+rather than by integer error codes.
+"""
+
+import errno
+import threading
+from . import exceptions
+from . import _error_translation as xlate
+from .bindings import libzfs_core
+from ._constants import MAXNAMELEN
+from ._nvlist import nvlist_in, nvlist_out
+
+
+def lzc_create(name, is_zvol=False, props=None):
+    '''
+    Create a ZFS filesystem or a ZFS volume ("zvol").
+
+    :param bytes name: a name of the dataset to be created.
+    :param bool is_zvol: whether to create a zvol (false by default).
+    :param props: a `dict` of ZFS dataset property name-value pairs (empty by default).
+    :type props: dict of bytes:Any
+
+    :raises FilesystemExists: if a dataset with the given name already exists.
+    :raises ParentNotFound: if a parent dataset of the requested dataset does not exist.
+    :raises PropertyInvalid: if one or more of the specified properties is invalid
+                             or has an invalid type or value.
+
+    Wraps ``int lzc_create(const char *fsname, dmu_objset_type_t type, nvlist_t *props)``.
+    '''
+    if props is None:
+        props = {}
+    if is_zvol:
+        ds_type = _lib.DMU_OST_ZVOL
+    else:
+        ds_type = _lib.DMU_OST_ZFS
+    nvlist = nvlist_in(props)
+    ret = _lib.lzc_create(name, ds_type, nvlist)
+    xlate.lzc_create_xlate_error(ret, name, is_zvol, props)
+
+
+def lzc_clone(name, origin, props=None):
+    '''
+    Clone a ZFS filesystem or a ZFS volume ("zvol") from a given snapshot.
+
+    :param bytes name: a name of the dataset to be created.
+    :param bytes origin: a name of the origin snapshot.
+    :param props: a `dict` of ZFS dataset property name-value pairs (empty by default).
+    :type props: dict of bytes:Any
+
+    :raises FilesystemExists: if a dataset with the given name already exists.
```

Comment:
I found that this can also raise `NameInvalid` while experimenting with the function.  And from inspecting the code, I see it could also raise `NameTooLong`.

I also initially found it confusing that the exception for passing a dataset name instead of a snapshot name is `NameInvalid`.  It would have been much more clear if the exception had told me the origin must be a snapshot.